Begin process of removing singleton use by StringMatcher

ggreenway opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: Begin process of removing singleton use by StringMatcher

This PR temporarily adds two versions of StringMatcherImpl: one that uses the existing singletons, and one that takes a factory context. 

Subsequent PRs will gradually convert all uses of StringMatcherImpl to use the factory context. Then the version using singletons will be deleted, along with the code that sets the singletons.
Additional Description: This work is being broken into multiple PRs because otherwise it would be very large and difficult to review.

Part of #32792
Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

ggreenway commented:
/retest

RyanTheOptimist reviewed (commented):
Looks reasonable, but I think ASAN is finding some correctness issues, eg:

source/common/common/matchers.h:104:66: runtime error: reference binding to null pointer of type 'Regex::Engine'

ggreenway commented:
> source/common/common/matchers.h:104:66: runtime error: reference binding to null pointer of type 'Regex::Engine'

Yep, some instances where I converted a nullptr to a reference. The reference was never used, but still gets flagged by asan. I think I've worked around it now, but waiting on CI. 

There was also a real failure on mobile that I think I've fixed.

RyanTheOptimist reviewed (commented):
LGTM, modulo coverage. But maybe that's hard to fix?

ggreenway commented on test/per_file_coverage.sh:
```diff
@@ -63,7 +63,7 @@ declare -a KNOWN_LOW_COVERAGE=(
 "source/extensions/listener_managers/validation_listener_manager:70.5"
 "source/extensions/watchdog/profile_action:83.3"
 "source/server:91.0" # flaky: be careful adjusting. See https://github.com/envoyproxy/envoy/issues/15239
-"source/server/config_validation:89.2"
```
I added `regexEngine()` to the `Server::Instance`. This isn't used (yet) in the config_validation path, because most of the uses of `StringMatcherImpl` still don't use it. Future PRs will convert all of those, and this line will get hit in coverage at that time.

ggreenway commented:
/retest
